velox/common/memory/MemoryAllocator.h

```diff
@@ -193,6 +211,26 @@ class MemoryAllocator : public std::enable_shared_from_this<MemoryAllocator> {
   /// the kind of the delegated memory allocator underneath.
   virtual Kind kind() const = 0;
 
+  /// Registers a 'Cache' that is used for freeing up space when this allocator
+  /// is under memory pressure. The allocator of registered 'Cache' needs to be
+  /// the same as 'this'.
+  void registerCache(Cache* cache) {
+    VELOX_CHECK_NULL(cache_);
+    VELOX_CHECK_NOT_NULL(cache);
+    VELOX_CHECK(cache->allocator() == this);
+    cache_ = cache;
+  }
+
+  /// Unregisters the 'Cache' that was previously registered. It is the caller's
+  /// responsibility to unregister the cache upon destruction of the 'Cache'.
+  ///
+  /// NOTE: Caller should make sure that unregistration of 'Cache' shall happen
```

Comment:
```
NOTE: the caller should ensures that unregistration of 'Cache' shall happen before its destruction and all the allocated cached memory has been freed.
```

It might not be safe to unregister the cache as the memory allocator can access the cache at any point so there is a race condition there. I don't think we really need unregister cache operation. Let's hold a shared reference on the cache from the allocator. And make sure there is no memory usage from cache when memory allocator destroys. We could add the allocated memory in cache interface.